projects/lpc40xx_freertos/l3_drivers/sources/uart_lab.c

```diff
@@ -0,0 +1,169 @@
+    const char byte = LPC_UART2->RBR;
+    xQueueSendFromISR(uart_rx_queue, &byte, NULL);
+    // fprintf(stderr, "I am in ISR, Just put %X data in queue\n", byte);
+  }
+}
+
+/**
+ * UART3 receiver interrupt handler fucntion
+ */
+static void uart3_receive_interrupt(void) {
+  const uint8_t interrupt_id_mask = (0b111);
+  const uint8_t interrupt_pending_bit = (1 << 0);
+  uint8_t interrupt_id;
+  // Check whether there is pending interrupt corresponding to UART3
+  if (!(LPC_UART3->IIR & interrupt_pending_bit))
+    interrupt_id = ((LPC_UART3->IIR >> 1) & interrupt_id_mask);
+  else {
+    fprintf(stderr, "UART Interrupt occured due to unconfigured UART cannel, habe look into it");
+    assert(0);
+  }
+
+  if ((interrupt_id == 2) && (LPC_UART3->LSR & lsr_reg_received_data_ready_bit_mask)) {
+    const char byte = LPC_UART3->RBR;
+    xQueueSendFromISR(uart_rx_queue, &byte, NULL);
+    // printf("I am in ISR, Just put %X data in queue\n", byte);
+  }
+}
+#endif
+
+/*******************************************************************************
+ *
+ *                      P U B L I C    F U N C T I O N S
+ *
+ ******************************************************************************/
+
+void uart_lab__init(uart_number_e uart, uint32_t peripheral_clock, uint32_t baud_rate) {
+  // Power on the peripheral
+  uint32_t uart_pconp_mask;
+  if (uart == UART_2)
+    uart_pconp_mask = (0x1 << 24);
+  else
+    uart_pconp_mask = (0x1 << 25);
+  LPC_SC->PCONP |= uart_pconp_mask;
+
+  lpc_reg *uart_reg = uarts[uart];
+
+  // Setup the Baud rate
+  // UART_baud_rate = (PCLK / (16 * ((DLM << 8) + DLL)))
+  set_dlab_bit(uart);
+  uint16_t DLs = peripheral_clock / (16 * baud_rate);
+  uart_reg->DLL |= (DLs & 0xFF);
+  uart_reg->DLM |= (DLs & (0xFF << 8));
+  clear_dlab_bit(uart);
+
+  // Configure LCR register
+  const uint8_t word_length_bit_mask = (3 << 0);
+  const uint8_t word_length_8bits_value_mask = (3 << 0);
+  uart_reg->LCR &= ~(word_length_bit_mask);
+  uart_reg->LCR |= (word_length_8bits_value_mask);
+
+  // Select UART pins and modes using IOCON
+  if (uart == UART_2) {
+    gpio__construct_with_function(GPIO__PORT_2, 8, GPIO__FUNCTION_2);
+    gpio__construct_with_function(GPIO__PORT_2, 9, GPIO__FUNCTION_2);
+
+  } else {
+    gpio__construct_with_function(GPIO__PORT_4, 28, GPIO__FUNCTION_2);
+    gpio__construct_with_function(GPIO__PORT_4, 29, GPIO__FUNCTION_2);
+  }
+}
+
+bool uart_lab__polled_get(uart_number_e uart, char *input_byte) {
+  clear_dlab_bit(uart);
+
+  if (uarts[uart]->LSR & lsr_reg_received_data_ready_bit_mask) {
+    *input_byte = uarts[uart]->RBR;
+    return true;
+  }
+  return false;
+}
+
+bool uart_lab__polled_put(uart_number_e uart, char output_byte) {
+  clear_dlab_bit(uart);
+  const uint8_t lsr_reg_transmit_buffer_empty_bit_mask = (1 << 5);
+  if (uarts[uart]->LSR & lsr_reg_transmit_buffer_empty_bit_mask) {
+    uarts[uart]->THR |= output_byte;
```

Comment:
@Vidushi1790  This line will get executed when the THR register is empty (based on if check above.). 
Below line 142, 'while check' is for polling THR register's data to get transmitted.